modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3647,6 +3652,82 @@ def apply_func(df):  # pragma: no cover
             func=apply_func,
         )
 
+        # no need aligning columns if there's only one row partition
+        if align_result_columns and result._partitions.shape[0] > 1:
+            # FIXME: the current reshuffling implementation guarantees us that there's only one column
+            # partition in the result, so we should never hit this exception for now, however
+            # in the future, we might want to make this implementation more broader
+            if result._partitions.shape[1] > 1:
+                raise NotImplementedError(
+                    "Aligning columns is not yet implemented for multiple column partitions."
+                )
+
+            # There're two implementations:
+            #   1. The first one work faster, but may stress the network a lot in cluster mode since
+            #      it gathers all the dataframes in a single ray-kernel.
+            #   2. The second one works slower, but only gathers light pandas.Index objects,
+            #      so there should be less stress on the network.
+            if not IsRayCluster.get():
+
+                def compute_aligned_columns(*dfs):
+                    """Take row partitions, filter empty ones, and return joined columns for them."""
+                    non_empty_dfs = [df for df in dfs if not df.empty]
```

Comment:
Empty dataframes can have columns, don't we need to account for them?

```python
pd.DataFrame([], columns=["a"]).empty
```

Reply:
Right, we only want to filter out those dfs that were empty originally and not those that produced an empty result, fixed that by adding the `skip_on_aligning` flag at the groupby kernel.

We have to filter out those frames that were empty originally, because they produce incorrect columns that shouldn't be in the result at all:

```python
>>> df
   a  b
0  1  3
1  1  4
2  2  5
3  2  6
>>> def func(group):
...     return (
...         pandas.DataFrame([[1, 2]], columns=["new1", "new2"])
...         if group.iloc[0, 0] == 1 else 
...         pandas.DataFrame(columns=["new2", "new3", "new4"]) # here we're producing an empty result
...     )
...
>>> df.groupby("a").apply(func) 
     new1 new2 new3 new4 # the columns of the empty result are joined to the non-empty result
1 0   1.0    2  NaN  NaN
>>> df.iloc[:0].groupby("a").apply(func) # runing groupby on an empty frame
Empty DataFrame
Columns: [a, b] # it produces columns that shouldn't be in the resulted frame, so we have to skip it when aligning
Index: []
```